### PR TITLE
Enable x86 CPU vectorization on windows [submodule sleef]

### PR DESCRIPTION
**Pull request by xuhancn**
Enable VEC on Windows OS.
1. Fix some type defination gap between Windows and Linux.
2. Fix some operator not support on Windows, such as [], /.
3. Enable static sleef library build on Windows.
4. Disable unsupported function overloading on MSVC.
5. Upgrade submodule sleef lib, which fixed build issue on Windows.
6. Fixed bazel build issues.
7. Fix test app not link to sleef on Windows.

Note: If rebuild fail after pulled this PR, please sync `sleef` submodule by run:
```cmd
git submodule sync
git submodule update --init --recursive
```

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @frank-wei

### CODE REVIEW

File: aten/src/ATen/native/cpu/GridSamplerKernel.cpp
Comment (jgong5) [inline aten/src/ATen/native/cpu/GridSamplerKernel.cpp:589]: no else?

File: aten/src/ATen/native/cpu/GridSamplerKernel.cpp
Comment (jgong5) [inline aten/src/ATen/native/cpu/GridSamplerKernel.cpp:696]: ditto

File: aten/src/ATen/native/cpu/GridSamplerKernel.cpp
Comment (jgong5) [inline aten/src/ATen/native/cpu/GridSamplerKernel.cpp:771]: ditto

File: aten/src/ATen/native/cpu/GridSamplerKernel.cpp
Comment (jgong5) [inline aten/src/ATen/native/cpu/GridSamplerKernel.cpp:1151]: ditto

File: aten/src/ATen/native/cpu/GridSamplerKernel.cpp
Comment (xuhancn) [reply aten/src/ATen/native/cpu/GridSamplerKernel.cpp]: It is WA during development, and fixed now.

**Review by jgong5 — APPROVED** [review]

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Do you think the test_grad_scaling_unscale_cpu_float32 failure is real?

File: caffe2/CMakeLists.txt
Comment (malfet) [inline caffe2/CMakeLists.txt:1529]: This should be guarded behind `USE_SLEEF`

Comment (malfet) [reply caffe2/CMakeLists.txt:1529]: Also, do you mind moving this to (and just remove MSVC part) https://github.com/pytorch/pytorch/blob/5410385c427cf2969da4d08b8ea9e4ed4c6b82d5/aten/src/ATen/CMakeLists.txt#L423

**Review by malfet — CHANGES_REQUESTED** [review]
Overall LGTM, but please re-use as much cmake code as possible

**Comment by xuhancn** [comment]
> Do you think the test_grad_scaling_unscale_cpu_float32 failure is real?

image
In my dev machine, this test is passed. 

And I trigger a dummy PR to check the origin case, but no Windows test was run. https://github.com/pytorch/pytorch/pull/119270

I guess it is not a real failure.

File: caffe2/CMakeLists.txt
Comment (xuhancn) [reply caffe2/CMakeLists.txt:1529]: > This should be guarded behind `USE_SLEEF`

1. USE_SLEEF is CPP macro: https://github.com/pytorch/pytorch/blob/8c2fde1fcf17a1108557740d581e1808cc5b7d38/aten/src/ATen/cpu/vec/vec256/vec256_float_neon.h#L38-L42
2. CMAKE not contains USE_SLEEF switch. Sleef is hard dependency of torch_cpu, if vec enabled. The only option is whether use system sleef: https://github.com/pytorch/pytorch/blob/8c2fde1fcf17a1108557740d581e1808cc5b7d38/CMakeLists.txt#L424

**Comment by xuhancn** [comment]
> > Do you think the test_grad_scaling_unscale_cpu_float32 failure is real?
> 
> image In my dev machine, this test is passed.
> 
> And I trigger a dummy PR to check the origin case, but no Windows test was run. #119270
> 
> I guess it is not a real failure.

Installed hypothesis, it passed also.
image

File: caffe2/CMakeLists.txt
Comment (xuhancn) [reply caffe2/CMakeLists.txt:1529]: Image
I referenced to the "ONNX", which is linked to torch_cpu success. It was processed as submodule in caffee2/CMakeLists.txt.
So I think I can only process sleef here.

**Review by malfet — APPROVED** [review]
Lets land it to avoid code getting stale and do some build followups later

**Comment by huydhn** [comment]
Only debug build is failing from what I see

**Comment by xuhancn** [comment]
It depends on https://github.com/shibatch/sleef/pull/510.

**Comment by xuhancn** [comment]
Sleef is suspend to merge new PR due to release: https://github.com/shibatch/sleef/discussions/498

**Comment by xuhancn** [comment]
Status update:
1. We still pending on sleef PR can't merge: https://github.com/shibatch/sleef/pull/510 , maintainer said that is limited bandwidth to review and merge code: https://github.com/shibatch/sleef/discussions/498#discussioncomment-8527530
2. I had to fork a sleef repo to move on my develop work. The forked repo contains my new sleef PR.
3. Current PR almost works on my forked sleef.
4. Still two bazel fail, and they are should be related to original sleef. It is not diffcute issue.
5. This PR can boost performance on all Windows machine, I wish we can catch up pytorch 2.3 release.
CC: @jgong5 @malfet @ezyang @blapie

File: .gitmodules
Comment (ezyang) [inline .gitmodules:154]: @malfet idk if you have an opinion about this, but in the past when we've needed to temporarily fork the repo, we typically updated the url in the submodule itself. For example, https://github.com/pytorch/pytorch/pull/48120

File: third_party/sleef_fork
Comment (ezyang) [inline third_party/sleef_fork:1]: The commit history is a bit messy. For ease of understanding the change here, it would be best if this was (1) the exact version of sleef we were previously on + (2) one patch with all of the changes for Windows compatibility. We then can update sleef by rebasing the patch in question.

**Comment by ezyang** [comment]
@xuhancn for speed of integration, may I recommend splitting the C++ changes (which hopefully we can just merge immediately) from the build system changes + sleef submodule changes?

**Comment by xuhancn** [comment]
> @xuhancn for speed of integration, may I recommend splitting the C++ changes (which hopefully we can just merge immediately) from the build system changes + sleef submodule changes?

1. Need Meta approve the fork sleef repo.
2. C++ change will turn on sleef, and it can't build pass in Windows without sleef update.
3. Due to current sleef was two years ago version, and build options changed. Can't seprate sleef and C++.

File: .gitmodules
Comment (xuhancn) [reply .gitmodules]: Done, for test.

File: third_party/sleef_fork
Comment (xuhancn) [reply third_party/sleef_fork:1]: rebase to one commit for code.

**Comment by xuhancn** [comment]
@ezyang  Current status:
1. Rebased commits into two commit: cpp related and sleef related.
2. Switch sleef to temp repo via URL.
3. Fixed bazel build fail.
This PR can pass all preCI test now.

Due to this PR will switch sleef repo to my temp repo, and it contains the necessary PR(https://github.com/shibatch/sleef/pull/510) is not merged.
Whether merge this PR, it up to Meta engineers. @ezyang @malfet

**Comment by xuhancn** [comment]
Add a locked pytorch dedicate branch: https://github.com/xuhancn/sleef/tree/pytorch_release

**Comment by xuhancn** [comment]
SLEEF will merge this PR(https://github.com/shibatch/sleef/pull/531) to fix the build issue. Let's wait for a while.

**Comment by blapie** [comment]
Have you tried with the most recent version of SLEEF? 
The .so disappeared due to the previous fix but got put back in this commit https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b

**Comment by xuhancn** [comment]
> Have you tried with the most recent version of SLEEF? The .so disappeared due to the previous fix but got put back in this commit [shibatch/sleef@60e76d2](https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b)

I will try latest sleef.

**Comment by xuhancn** [comment]
> Have you tried with the most recent version of SLEEF? The .so disappeared due to the previous fix but got put back in this commit [shibatch/sleef@60e76d2](https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b)

@zou3519 retry to latest sleef. It should fixed your problem.

**Comment by zou3519** [comment]
> > Have you tried with the most recent version of SLEEF? The .so disappeared due to the previous fix but got put back in this commit [shibatch/sleef@60e76d2](https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b)
> 
> @zou3519 retry to latest sleef. It should fixed your problem.

How does one "retry to latest sleef"? The thing I do when installing PyTorch is checking out all submodules. Is that sufficient or do I need to install something new?

**Comment by xuhancn** [comment]
> > > Have you tried with the most recent version of SLEEF? The .so disappeared due to the previous fix but got put back in this commit [shibatch/sleef@60e76d2](https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b)
> > 
> > 
> > @zou3519 retry to latest sleef. It should fixed your problem.
> 
> How does one "retry to latest sleef"? The thing I do when installing PyTorch is checking out all submodules. Is that sufficient or do I need to install something new?

Sleef seems some regression, The latest PR(https://github.com/shibatch/sleef/pull/535) looks like fix it.

**Comment by xuhancn** [comment]
@zou3519 ping you at slack, free to discuss?

**Comment by xuhancn** [comment]
@zou3519 may I known if you run the command update `sleef` submodule?
```cmd
git submodule sync
git submodule update --init --recursive
```

**Comment by xuhancn** [comment]
Hi @zou3519 here two possible reason caused you build fail:
1. Sleef regression, the new sleef PR(https://github.com/shibatch/sleef/pull/535) fixed it. I have update the sleef to the fixed version.
2. You didn't update `sleef` submodule, please run below cmd:
```cmd
git submodule sync
git submodule update --init --recursive
```
This PR passed all of post CI, I some confidence to its quality.

**Comment by xuhancn** [comment]
> @pytorchbot revert -m "This caused build failures for 2+ pytorch devs, so we're reverting it to be safe" -c weird

Please sync `sleef` submodule by run:
```cmd
git submodule sync
git submodule update --init --recursive
```

**Comment by xuhancn** [comment]
> Have you tried with the most recent version of SLEEF? The .so disappeared due to the previous fix but got put back in this commit [shibatch/sleef@60e76d2](https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b)

@blapie merge this PR again, with latest sleef.

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for revert your change one more time but the hard part is that it breaks lot of internal builds' -c ghfirst

@ezyang @malfet As reviewers, could one of you help take a look at the failures on D55166757 to unblock this?  From what I see, the build error is `unknown type name 'SLEEF_CONST'`.  Do we have guide somewhere on how to handle external module update in fbcode?

**Comment by xuhancn** [comment]
> @pytorchbot revert -m 'Sorry for revert your change one more time but the hard part is that it breaks lot of internal builds' -c ghfirst
> 
> @ezyang @malfet As reviewers, could one of you help take a look at the failures on D55166757 to unblock this? From what I see, the build error is `unknown type name 'SLEEF_CONST'`. Do we have guide somewhere on how to handle external module update in fbcode?

`SLEEF_CONST` Need upgrade to latest sleef lib. @huydhn original sleef can't working on pytorch Windows.

**Comment by xuhancn** [comment]
> > @pytorchbot revert -m 'Sorry for revert your change one more time but the hard part is that it breaks lot of internal builds' -c ghfirst
> > @ezyang @malfet As reviewers, could one of you help take a look at the failures on D55166757 to unblock this? From what I see, the build error is `unknown type name 'SLEEF_CONST'`. Do we have guide somewhere on how to handle external module update in fbcode?
> 
> `SLEEF_CONST` Need upgrade to latest sleef lib. @huydhn original sleef can't working on pytorch Windows.

@huydhn I have add some code to handle this issue: https://github.com/pytorch/pytorch/pull/118980/commits/aa83a5277b59651f75595de14bdaed8b3f4f2927 and re-merge it. Please help me on trace issue.

**Comment by huydhn** [comment]
Let me import your change first to see if it fixes the internal build failures, and I will merge this once I could confirm it

**Comment by huydhn** [comment]
Another build failure shows up FYI:

```
aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:492:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_logf8_u10);
```

**Comment by xuhancn** [comment]
> Another build failure shows up FYI:
> 
> ```
> aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:492:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
>     return map(Sleef_logf8_u10);
> ```

Please import with it: https://github.com/pytorch/pytorch/pull/118980/commits/da32e86c1083892ee6c9a023f9f673d73b1b092b

**Comment by xuhancn** [comment]
> @huydhn has imported this pull request. If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55208753).

@huydhn Any update?

**Comment by huydhn** [comment]
Sorry for the delay, @malfet mentions that sleef module is already there internall on fbcode, so I think I could try to update it to block this change.

**Comment by anijain2305** [comment]
Not sure if related. I am on `main` branch (not on this PR) and seeing bulid failure like these

```
/data/users/anijain/pytorch/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h: In instantiation of ‘at::vec::AVX2::Vectorized<T> at::vec::AVX2::Vectorized16<T>::cosh() const [with T = c10::BFloat16]’:
/data/users/anijain/pytorch/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:376:3:   required from here
/data/users/anijain/pytorch/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:503:16: error: invalid conversion from ‘__m256 (*)(__m256)’ to ‘const __m256 (*)(__m256)’ [-fpermissive]
  503 |     return map(Sleef_coshf8_u10);
      |                ^~~~~~~~~~~~~~~~
      |                |
      |                __m256 (*)(__m256)
```

I have already sync'd the submodules. Any idea?

**Comment by xuhancn** [comment]
> Not sure if related. I am on `main` branch (not on this PR) and seeing bulid failure like these
> 
> ```
> /data/users/anijain/pytorch/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h: In instantiation of ‘at::vec::AVX2::Vectorized<T> at::vec::AVX2::Vectorized16<T>::cosh() const [with T = c10::BFloat16]’:
> /data/users/anijain/pytorch/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:376:3:   required from here
> /data/users/anijain/pytorch/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:503:16: error: invalid conversion from ‘__m256 (*)(__m256)’ to ‘const __m256 (*)(__m256)’ [-fpermissive]
>   503 |     return map(Sleef_coshf8_u10);
>       |                ^~~~~~~~~~~~~~~~
>       |                |
>       |                __m256 (*)(__m256)
> ```
> 
> I have already sync'd the submodules. Any idea?
> 
> update - I fixed it in this PR - #122723. Not sure though.

Please wait for a while, My(this PR) will fix the sleef build issue. And current blocked on fb_code.

**Comment by xuhancn** [comment]
merge PR avoid code getting stale, @huydhn please update fb_code sleef to https://github.com/shibatch/sleef/commit/60e76d2bce17d278b439d9da17177c8f957a9e9b, it should be works well. CC: @malfet

**Comment by atalman** [comment]
Please note multiple jobs are broken. here are some of the errors:
```
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:495:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_logf8_u10);
               ^~~~~~~~~~~~~~~
xplat/caffe2/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:157:52: note: in instantiation of member function 'at::vec::Vectorized16<c10::BFloat16>::log' requested here
                return (x_vec / (kOneVec - x_vec)).log();
                                                   ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:281:58: note: passing argument to parameter 'vop' here
  Vectorized<T> map(SLEEF_CONST __m256 (*SLEEF_CONST_OLD vop)(__m256)) const {
                                                         ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:495:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_logf8_u10);
               ^~~~~~~~~~~~~~~
xplat/caffe2/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:157:52: note: in instantiation of member function 'at::vec::Vectorized16<c10::Half>::log' requested here
                return (x_vec / (kOneVec - x_vec)).log();
                                                   ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:281:58: note: passing argument to parameter 'vop' here
  Vectorized<T> map(SLEEF_CONST __m256 (*SLEEF_CONST_OLD vop)(__m256)) const {
                                                         ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:510:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_sinhf8_u10);
               ^~~~~~~~~~~~~~~~
xplat/caffe2/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:371:60: note: in instantiation of member function 'at::vec::Vectorized16<c10::BFloat16>::sinh' requested here
        [=](Vectorized<scalar_t> self_vec){return self_vec.sinh();});
                                                           ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:281:58: note: passing argument to parameter 'vop' here
  Vectorized<T> map(SLEEF_CONST __m256 (*SLEEF_CONST_OLD vop)(__m256)) const {
                                                         ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:510:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_sinhf8_u10);
               ^~~~~~~~~~~~~~~~
xplat/caffe2/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:371:60: note: in instantiation of member function 'at::vec::Vectorized16<c10::Half>::sinh' requested here
        [=](Vectorized<scalar_t> self_vec){return self_vec.sinh();});
                                                           ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:281:58: note: passing argument to parameter 'vop' here
  Vectorized<T> map(SLEEF_CONST __m256 (*SLEEF_CONST_OLD vop)(__m256)) const {
                                                         ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:516:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_coshf8_u10);
               ^~~~~~~~~~~~~~~~
xplat/caffe2/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:380:60: note: in instantiation of member function 'at::vec::Vectorized16<c10::BFloat16>::cosh' requested here
        [=](Vectorized<scalar_t> self_vec){return self_vec.cosh();});
                                                           ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:281:58: note: passing argument to parameter 'vop' here
  Vectorized<T> map(SLEEF_CONST __m256 (*SLEEF_CONST_OLD vop)(__m256)) const {
                                                         ^
xplat/caffe2/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:516:16: error: cannot initialize a parameter of type '__m256 (*)(__m256)' with an lvalue of type 'const __m256 (__m256)': different return type ('__m256' (vector of 8 'float' values) vs 'const __m256' (vector of 8 'float' values))
    return map(Sleef_coshf8_u10);
               ^~~~~~~~~~~~~~~~
xplat/caffe2/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:380:60: note: in instantiation of member function 'at::vec::Vectorized16<c10::Half>::cosh' requested here
        [=](Vectorized<scalar_t> self_vec){return self_vec.cosh();});
```

**Comment by huydhn** [comment]
@atalman I'm importing the sleef module to unblocking this internally, I have the diff to import sleef at D55567582.  We can rebase this on top to see if it fixes the build issue.

**Pull request closed without merging**
